### PR TITLE
Add file size to DirectoryEntry

### PR DESCRIPTION
**Pull request by ravenblackx**
Commit Message: Add file size to DirectoryEntry
Additional Description: Directory [find] iterator is potentially useful for file system cache and, in future, serving content directly from file system. The iterator already has access to file size from `stat()` or its Windows equivalent, but it is not accessible to the caller. Using the current available APIs, the caller would have to make an additional call to `stat` for each file they wish to know the size of. This change exposes the file size through the existing API, sparing those additional OS calls in the case that the size is useful, at the cost of a small value-copy.
In passing, fixes a mistaken use of `errno` that should have been the error number from the os-call's returned structure, and updates a throw-test to not require a nominally private function to be public.
Risk Level: Low. Simple change, existing fields should be unchanged.
Testing: Tests updated to match new values, one test added to validate that file size is returned.
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: Updated both win32 and posix versions.

### CODE REVIEW

File: envoy/filesystem/filesystem.h
Comment (jmarantz) [inline envoy/filesystem/filesystem.h:172]: nit: I don't feel a compelling need to document this variable any more than it's well-chosen name :) 

No worries if you prefer to leave it though.

File: source/common/filesystem/posix/directory_iterator_impl.cc
Comment (jmarantz) [inline source/common/filesystem/posix/directory_iterator_impl.cc:49]: since you are computing full_path in makeEntry you can drop the 2 temps above.

File: source/common/filesystem/posix/directory_iterator_impl.cc
Comment (jmarantz) [inline source/common/filesystem/posix/directory_iterator_impl.cc:75]: pre-existing, but should we assign file_type in the 'else' or throw an exception or something?

File: source/common/filesystem/posix/directory_iterator_impl.cc
Comment (jmarantz) [inline source/common/filesystem/posix/directory_iterator_impl.cc:70]: pre-existing, but can makeEntry be called during a request? It seems like it can, in which case we should switch from throwing an exception to returning an absl::StatusOr. It is not permitted to throw exceptions on worker threads in Envoy.

Actually throwing them is OK if they crash the program (for some definition of OK). But we disallow *catch* in worker threads.

This is beyond the scope of this PR but would it be worth it to throw in a TODO while you are here?

File: test/common/filesystem/directory_test.cc
Comment (jmarantz) [inline test/common/filesystem/directory_test.cc:55]: should we check for errors here, even with (say) an EXPECT_* macro?

**Review by jmarantz — COMMENTED** [review]
lgtm with a few nits.

File: envoy/filesystem/filesystem.h
Comment (ravenblackx) [reply envoy/filesystem/filesystem.h:172]: I'd rather leave it because I don't know what to do with the spacing otherwise, haha. A blank line between variables is weird, but with no blank line the comment above the *other* variable seems to cover both.

File: source/common/filesystem/posix/directory_iterator_impl.cc
Comment (ravenblackx) [reply source/common/filesystem/posix/directory_iterator_impl.cc:47]: Done.

File: source/common/filesystem/posix/directory_iterator_impl.cc
Comment (ravenblackx) [reply source/common/filesystem/posix/directory_iterator_impl.cc:71]: Added TODO. (I too thought throwing an exception was a weird choice here.)

File: source/common/filesystem/posix/directory_iterator_impl.cc
Comment (ravenblackx) [reply source/common/filesystem/posix/directory_iterator_impl.cc:76]: It's assigned at construct-time (because it's used both for the 'else' and for the condition in the first if block). Added an else-comment to clarify that the fallthrough is intentional.

File: test/common/filesystem/directory_test.cc
Comment (ravenblackx) [reply test/common/filesystem/directory_test.cc:59]: Added test-asserts, and one in the pre-existing similar spot too.

File: envoy/filesystem/filesystem.h
Comment (jmarantz) [inline envoy/filesystem/filesystem.h:173]: should we include size_bytes_ in the operator== impl? Actually it is not obvious why you previously needed one of those. Would the default operator not work?

WIth this explicit definition, now it will be different of course as this one ignores the size.

Also, what is this operator== used for?

**Review by jmarantz — COMMENTED** [review]
lgtm modulo minor remaining nits.

File: envoy/filesystem/filesystem.h
Comment (ravenblackx) [reply envoy/filesystem/filesystem.h:173]: Yikes, good catch. The == is at least used by the test, so fixing it revealed that I wasn't testing what I thought I was testing. It also seems to be used in a != operator for the iterator, for the `!= end()` comparison, which didn't really care about the size part. Fixed.

Comment (ravenblackx) [reply envoy/filesystem/filesystem.h:173]: Oh, and there is no default == operator apparently. I too thought there was for pod types, but apparently not until C++20 and even then you still have to specify it with `=default`. I guess we've both been mixing languages too much!

Comment (ravenblackx) [reply envoy/filesystem/filesystem.h:173]: I *think* keeping the `==` (but fixed) is the least-confusing option for this. It's unfortunate that iterators use `!=` as the 'is at the end' check. We *could* override that to be specifically a test for end-iterator but that would be way more confusing than having an otherwise-unnecessary == operator.

And having the == operator be only as specific as it needs to be for iterator end detection would also just be more confusing than helpful, and would require rewriting the tests significantly. So I think what we have now (a handwritten default == operator) is the least horrible option.

File: test/common/filesystem/directory_test.cc
Comment (jmarantz) [inline test/common/filesystem/directory_test.cc:53]: nit: any reason not to use the filesystem itself here, rather than dropping to iostream? It's got an API for returning the error text which would be nice here.

up to you though, I'm OK leaving it like this, but generally it ruffles my feathers to swallow error messages like this.

Comment (ravenblackx) [reply test/common/filesystem/directory_test.cc:54]: I guess depending on an adjacent thing for the unit tests of a thing is a little dubious? The main reason I did it this way is because it was already done this way and I'm trying to not change everything I look at and think yuck about, because that would take [even more] forever.

Comment (jmarantz) [reply test/common/filesystem/directory_test.cc:54]: I'm not sure why ... that's part of the subsystem under test. 

But I also don't feel that strongly about it.

**Review by jmarantz — APPROVED** [review]
/assign-from envoyproxy/senior-maintainers

File: envoy/filesystem/filesystem.h
Comment (jmarantz) [reply envoy/filesystem/filesystem.h:174]: sounds good

**Review by jmarantz — APPROVED** [review]

File: envoy/filesystem/filesystem.h
Comment (snowp) [inline envoy/filesystem/filesystem.h:174]: Any way we can make this less of a footgun? One thing that comes to mind is wrapping it in a call that ASSERTs that `type_` is `Regular`? Or perhaps making it a absl::optional?

**Review by snowp — DISMISSED** [review]
Overall lgtm, just one suggestion

File: envoy/filesystem/filesystem.h
Comment (ravenblackx) [reply envoy/filesystem/filesystem.h:174]: Asserting that `type_` is `Regular` doesn't actually solve the problem, because on Windows with a symlink the type is regular and the size is unknown (or maybe it's the size of the linked path, I forget - either way, not the size of the file).

Using `absl::optional` is a good idea though, to make it a hard failure instead of a quiet failure if size is used when it shouldn't be.

File: source/common/filesystem/win32/directory_iterator_impl.cc
Comment (alyssawilk) [inline source/common/filesystem/win32/directory_iterator_impl.cc:46]: should this all be in the final else?

File: test/common/filesystem/directory_test.cc
Comment (alyssawilk) [inline test/common/filesystem/directory_test.cc:19]: sorry if I'm missing something obvious but why NOLINTNEXTLINE? Can you not just have Envoy standard naming?

File: test/common/filesystem/directory_test.cc
Comment (alyssawilk) [inline test/common/filesystem/directory_test.cc:72]: sorry ignorance of ostream but if you fail to write will file become null/false?

File: envoy/filesystem/filesystem.h
Comment (alyssawilk) [inline envoy/filesystem/filesystem.h:178]: sorry if I'm missing but did we test this new equivalence?

**Review by alyssawilk — COMMENTED** [review]
Looks good overall!
/wait

File: envoy/filesystem/filesystem.h
Comment (ravenblackx) [reply envoy/filesystem/filesystem.h:178]: We did now.

**Pull request merged**
